probe-rs/src/architecture/riscv/communication_interface.rs

```diff
@@ -0,0 +1,848 @@
+
+        let response_value: u128 = response_bytes.iter().enumerate().fold(0, |acc, elem| {
+            let (byte_offset, value) = elem;
+            acc + ((*value as u128) << (8 * byte_offset))
+        });
+
+        // Verify that the transfer was ok
+
+        let op = (response_value & 0x3) as u8;
+
+        if op != 0 {
+            return Err(RiscvError::DmiTransfer(
+                DmiOperationStatus::parse(op).unwrap(),
+            ));
+        }
+
+        let value = (response_value >> 2) as u32;
+
+        Ok(value)
+    }
+
+    pub(crate) fn read_dm_register<R: DebugRegister>(&mut self) -> Result<R, RiscvError> {
+        log::debug!("Reading DM register '{}' at {:#010x}", R::NAME, R::ADDRESS);
+
+        // TODO: parameter for this
+        let retry_duration = Duration::from_secs(5);
+
+        let start_time = Instant::now();
+
+        loop {
+            // Prepare the read, answer will be read back in a second operation
+            match self.dmi_register_access(R::ADDRESS as u64, 0, DmiOperation::Read) {
+                Ok(_) => break,
+                Err(RiscvError::DmiTransfer(DmiOperationStatus::RequestInProgress)) => {
+                    // Operation still in progress, reset dmi status and try again.
+                    self.dmi_reset()?;
+                }
+                Err(e) => return Err(e),
+            }
+
+            if start_time.elapsed() > retry_duration {
+                return Err(RiscvError::Timeout);
+            }
+        }
+
+        let response;
+
+        loop {
+            // Perform the actual read
+            match self.dmi_register_access(0, 0, DmiOperation::NoOp) {
+                Ok(r) => {
+                    response = r;
+                    break;
+                }
+                Err(RiscvError::DmiTransfer(DmiOperationStatus::RequestInProgress)) => {
+                    // Operation still in progress, reset dmi status and try again.
+                    self.dmi_reset()?;
+                }
+                Err(e) => return Err(e),
+            }
+
+            if start_time.elapsed() > retry_duration {
+                return Err(RiscvError::Timeout);
+            }
+        }
+
+        //let response = self.dmi_register_access(0, 0, DmiOperation::Read)?;
+
+        log::debug!(
+            "Read DM register '{}' at {:#010x} = {:#010x}",
+            R::NAME,
+            R::ADDRESS,
+            response
+        );
+
+        Ok(response.into())
+    }
+
+    pub(crate) fn write_dm_register<R: DebugRegister>(
+        &mut self,
+        register: R,
+    ) -> Result<(), RiscvError> {
+        // write write command to dmi register
+
+        let data = register.into();
+
+        log::debug!(
+            "Write DM register '{}' at {:#010x} = {:#010x}",
+            R::NAME,
+            R::ADDRESS,
+            data
+        );
+
+        self.dmi_register_access(R::ADDRESS as u64, data, DmiOperation::Write)?;
```

Comment:
Handle dmi busy.

Reply:
How would you handle it? Currently, an error is returned if the dmi was already busy. Would you retry with a timeout?

Reply:
I would clear the error with a soft reset, and retry again for a limited amount of time.

On the GD32V I had cases, where it would suddenly get stuck here. I fixed it by increasing the idle time to > the recommended one, and later by retrying after very short timeout.

Reply:
Sounds good, I've added a timeout here as well, and now it's handled the same as when reading a `dm` register.